designs/_posts/2017-07-13-improved-command-line-reporting.md

```diff
@@ -0,0 +1,540 @@
+       invocation. For example,
+
+        1. a "semantic" command line that includes only flags that change what 
+           the correct build is.
+        1. a "behavioral" command line with flags that define Bazel's 
+           constraints about how to get there (jvm args, logging flags, 
+           sandboxing, etc.)
+
+	+  As there might be many possible filters we want to apply, we shouldn't 
+       try and guess what users will find useful, but provide the flag category
+       information with the rest of the command line, with command lines 2 and 
+       3. The user/tool can then filter them as needed.
+  
+Note that none of these claims to be a "reproducible" command line. As stated 
+and explained above, that is not a goal of this effort. Some users would just 
+want the original command line to copy paste back into their terminal, which 
+could be type 1 (tool) or 2 (literal), or want the "equivalent" command line, 
+type 3. Type 4 becomes relevant if they want to isolate the settings which 
+caused a problem, filtering out the noise. 
+
+## Work Required
+
+### Overview
+
+There are a fair number of isolated changes that need to be made before we can
+provide new structured command lines as part of Bazel's output. The following
+can be done in parallel:
+
++  To get command line 1, Bazel-invoking tools need to be able to pass Bazel 
+   their command lines.
++  Refactoring the OptionsParser to maintain correct information about the
+   initial and final states of the command line.
++  Passing the client arguments from the client to the server literally.
++  Adding the command line reporting structure to the Build Event Protocol 
+   (BEP).
++  To filter command lines for command lines 4, we need better option 
+   categorization. 
+
+Tying everything together and reporting the command lines via the build event 
+stream would need to come last.
+
+Details for these are below.
+
+
+
+
+### Passing the Bazel-invoking tool's command line to the server
+
+A `--summoning_command` command option will serve to accept the original tool's
+command line. It will accept one of two types of values: [1] a compiled proto 
+containing the list of strings that represents the calling executable's argv[],
+or [2] a single string. 
+
+This can quickly become an unbounded dance - "what about (the tool that 
+invoked)+ Bazel?" We are intentionally not going to support this. Flags are 
+not a great tool for extended information passing, and while there are ways to 
+do it, it raises questions of ordering and correctness that we will not address 
+here. Bazel will accept 1 value to this flag, for the purpose of helping users 
+understand the origin of the output, not to keep an unbounded log of everything 
+that ever led to an invocation. If additional logging is needed, it might be 
+worth adding state somewhere outside of a flag value.
+
+#### Alternative formats considered
+
++  Passing by command arg
+
+    +  Compiled proto (Format [1])
```

Comment:
The format footnotes don't appear to render in GitHub, not sure how they'll look on docs.bazel.build.